src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionPool.cs

```diff
@@ -132,6 +160,53 @@ internal sealed class HttpConnectionPool : IDisposable
             }
         }
 
+        public async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            while (true)
+            { 
+                // Loop on connection failures and retry if possible.
+
+                HttpConnection connection = await GetConnectionAsync(request, cancellationToken);
+
+                if (connection.IsNewConnection)
+                {
+                    return await connection.SendAsync(request, cancellationToken);
+                }
+
+                try
+                {
+                    return await connection.SendAsync(request, cancellationToken);
+                }
+                catch (HttpRequestException e)
+                {
+                    if (!(e.InnerException is IOException))
+                    {
+                        throw;
+                    }
+
+                    if (!connection.CanRetry)
+                    {
+                        throw;
+                    }
+                }
```

Comment:
I don't know that it matters at all, but this could also be:
```C#
catch (HttpRequestException e) when (e.InnerException is IOException && connection.CanRetry)
{
}
```
I personally find that easier to understand, but to each his own.